docker_cli/_utils.py

```diff
@@ -0,0 +1,256 @@
+
+    if configure:
+        init_lines: str = (
+            "RUN " + str(package_mgr.generate_configure_command(stringify=True)) + "\n"
+        )
+    else:
+        init_lines = ""
+
+    url_program, url_init = _url_reader_check(image=base, package_mgr=package_mgr)
+    init_lines += url_init
+
+    run(split(f"docker image rm {tag}"), stdout=DEVNULL, stderr=DEVNULL)
+
+    return package_mgr, url_program, init_lines
+
+
+def _parse_cuda_info(cuda_version: str) -> Tuple[int, int]:
+    """
+    Turns a cuda version string into a major and minor version.
+
+    Parameters
+    ----------
+    cuda_version : str
+        The version string.
+
+    Returns
+    -------
+    Tuple[int, int]
+        The major and minor versions, in that order
+
+    Raises
+    ------
+    ValueError
+        If the input string does not encode a valid CUDA version number.
+    """
+    cuda_ver_pattern = re.compile(r"^(?P<major>[0-9]+)\." r"(?P<minor>[0-9]+)$")
+    cuda_ver_match = re.match(cuda_ver_pattern, cuda_version)
+    if not cuda_ver_match:
+        raise ValueError(f"Malformed CUDA version: {cuda_version}")
+    cuda_ver_match_groups = cuda_ver_match.groupdict()
+
+    cuda_major: int = int(cuda_ver_match_groups["major"])
+    cuda_minor: int = int(cuda_ver_match_groups["minor"])
+
+    return (cuda_major, cuda_minor)
+
+
+def _is_conda_pkg_name(line: str) -> bool:
+    """
+    Returns if a line appears to be an anaconda package URL
+
+    Used for filtering lockfiles
+
+    Parameters
+    ----------
+    line : str
+        The line
+
+    Returns
+    -------
+    bool
+        True if the line appears to be an anaconda package URL, false otherwise
+    """
+    conda_pkg_pattern = re.compile(r"^https:\/\/conda.anaconda.org\/\S*$")
+    return re.match(conda_pkg_pattern, line) is not None
+
+
+def test_image(image: Image, expression: str) -> bool:
+    """
+    Runs a test expression on an image.
+
+    Parameters
+    ----------
+    image : Image
+        The image.
+    expression : str
+        The expression.
+
+    Returns
+    -------
+    bool
+        True if "test" returned with status 0, False otherwise.
+    """
+    try:
+        image.run(f"test -d {expression}")
+        return True
+    except CalledProcessError as err:
+        if err.returncode == 1:
+            return False
+        else:
+            raise
+
+
+class UniqueGenerator:
+    """Generates random, threadsafe strings of lowercase letters and digits."""
+
+    lock = Lock()
+
+    @staticmethod
+    def generate(k: int = 10) -> str:
+        """
+        Generates random, threadsafe strings of lowercase letters and digits.
+
+        Parameters
+        ----------
+        k : int, optional
+            The length of the string to be generated. Defaults to 10.
+
+        Returns
+        -------
+        str
+            The random string.
+        """
+        with UniqueGenerator.lock:
+            return "".join(random.choices(ascii_lowercase + digits, k=k))
```

Comment:
Lol I didn't realize that the `random` module was not threadsafe. Good to know.

FYI, there's a trick to doing this without having to create a `UniqueGenerator` class, so that the `generate()` method can just be a free-standing function.

The trick is to use an internal helper function with an optional argument. The default value of the argument will be created just once (when the python module is first loaded).

```python
# `generate()` is a free function -- not a class method.
def generate(k: int = 10) -> str:
    
    # `lock` is initialized just once (when the Python module is first loaded).
    # Every invocation of the function will use the same lock instance.
    def helper(lock: Lock = Lock()) -> str:
        with lock:
            return "".join(random.choices(ascii_lowercase + digits, k=k))

    return helper()
```

The internal `lock` is initialized just once and reused every time the function is called. If the function is called in multiple threads, each thread will use the same lock.

This trick is nice because the lock is now purely an implementation detail -- it doesn't affect the interface at all and users don't need to know about it.

Reply:
Done! An additional benefit to this method is that it removes the need for a shell class altogether - the only reason I needed that was to hold onto the lock, but since that will be held in the helper function, `generate_random_string()` can sit as a function instead of being unnecessarily object-oriented.